pycloudlib/lxd/instance.py

```diff
@@ -333,3 +334,85 @@ def wait_for_stop(self):
                 return
             time.sleep(1)
         raise TimeoutError
+
+
+class LXDVirtualMachineInstance(LXDInstance):
+    """LXD Virtual Machine backed instance."""
+
+    def __init__(self, name, key_pair=None, execute_via_ssh=True):
+        """Set up instance.
+
+        Args:
+            name: name of instance
+            key_pair: SSH key object
+        """
+        super().__init__(
+            name=name,
+            key_pair=key_pair,
+            execute_via_ssh=execute_via_ssh
+        )
+
+        self.series = self._extract_release_from_instance_info()
+
+    def _run_command(self, command, stdin):
+        """Run command in the instance."""
+        if self.execute_via_ssh:
+            return super()._run_command(command, stdin)
+
+        if self.series == "xenial":
+            msg = (
+                "The xenial instance does not support executing commads"
+                " with exec."
+            )
+            raise Exception(msg)
+
+        base_cmd = [
+            'lxc', 'exec', self.name, '--', 'sudo', '-u', self.username, '--'
+        ]
+        return subp(base_cmd + list(command), rcs=None)
+
+    def _lxc_instance_info(self) -> dict:
+        """Return a dict of the output of ``lxc info <instance_name>``.
+
+        Args:
+            instance_name: string, name of the LXD instance
+
+        Returns:
+            A dict produced by loading the YAML emitted by ``lxc info
+            <instance_name>``, or the empty dict if either the command
+            or YAML load fails.
+        """
+        raw_info = subp(["lxc", "info", self.name], rcs=())
+        if raw_info.ok:
+            try:
+                return yaml.safe_load(raw_info.replace("\t", " "))
+            except yaml.YAMLError:
+                pass
+        return {}
+
+    def _extract_release_from_instance_info(self):
+        """Extract the base release from the instance.
+
+        We will use the profile created for LXD virtual machines
+        to infer which release was used to create that instance
+
+        Returns:
+            A string containing the base release from the instance
+        """
+        instance_info = self._lxc_instance_info()
+        release = None
+        try:
+            profiles = str(instance_info["Profiles"])
+
+            pattern_search = re.search(
+                r"pycloudlib-vm-(?P<release>.*)",
```

Comment:
Once we fix #75, everything from focal and later should be sharing a profile; this determination will break.

Given that we know exactly the image ID that we're using in `_BaseLXD.launch`, we could pass that in as a parameter to the instance instantiation instead, and then call `self.cloud._extract_release_from_image_id` instead of performing this determination?

Reply:
This is true during a launch situation, but if we are using getting an existing instance through `cloud.get_instance` we would not have the `image ID` to get that information from. That's why I am extracting this from `lxc info` command.

But you are totally right that the regex will no longer work. I am fixing that now

Reply:
> This is true during a launch situation, but if we are using getting an existing instance through `cloud.get_instance` we would not have the `image ID` to get that information from. That's why I am extracting this from `lxc info` command.

If we're getting an existing instance through `cloud.get_instance`, there's no guarantee that it will have a pycloudlib profile applied to it, I think? So I think we're always going to have corner cases around `get_instance`.

If we don't handle this, what's the consequence? I think it will only mean that we don't emit a warning log message, once my comment above is taken into account.

Reply:
Yes, I agree that the `cloud.get_instance` will have corner cases if the user created the vm outside pycloudlib and are now trying to use in pycloudlib. However, what will happen in that situation is just that we not alert the user that `exec`  commands will not work in xenial vms.

The reason I made this is to try to be explicit about warning users of possible problems in the code. But maybe I am overthinking this. Specially now that we are just emitting a warning instead of raising an `Exception`. If you agree that this solution is too complicated to just alert the user about potential xenial problems, just let me know and I will drop it